src/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs

```diff
@@ -3,13 +3,90 @@
 
 using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
 using System.IO;
 using System.Threading.Tasks;
 
 namespace System.Net.Sockets
 {
     public partial class Socket
     {
+        partial void ValidateForMultiConnect(bool isMultiEndpoint)
+        {
+            // ValidateForMultiConnect is called before any {Begin}Connect{Async} call,
+            // regardless of whether it's targeting an endpoint with multiple addresses.
+            // If it is targeting such an endpoint, then any exposure of the socket's handle
+            // or configuration of the socket we haven't tracked would prevent us from
+            // replicating the socket's file descriptor appropriately.  Similarly, if it's
+            // only targeting a single address, but it already experienced a failure in a
+            // previous connect call, then this is logically part of a multi endpoint connect,
+            // and the same logic applies.  Either wya, in such a situation we throw.
+            if (_handle.ExposedHandleOrUntrackedConfiguration && (isMultiEndpoint || _handle.LastConnectFailed))
+            {
+                ThrowMultiConnectNotSupported();
+            }
+
+            // If the socket was already used for a failed connect attempt, replace it
+            // with a fresh one, copying over all of the state we've tracked.
+            ReplaceHandleIfNecessaryAfterFailedConnect();
+            Debug.Assert(!_handle.LastConnectFailed);
+        }
+
+        internal void ReplaceHandleIfNecessaryAfterFailedConnect()
+        {
+            if (!_handle.LastConnectFailed)
+            {
+                return;
+            }
+
+            // Copy out values from key options. The copied values should
+            // be kept in sync with the list in SafeCloseSocket.TrackOption.
+            bool broadcast = false, dontFragment = false, noDelay = false;
+            int receiveSize = -1, receiveTimeout = -1, sendSize = -1, sendTimeout = -1;
+            short ttl = -1;
+            LingerOption linger = null;
+            if (_handle.IsTrackedOption(TrackedSocketOptions.DontFragment)) dontFragment = DontFragment;
+            if (_handle.IsTrackedOption(TrackedSocketOptions.EnableBroadcast)) broadcast = EnableBroadcast;
+            if (_handle.IsTrackedOption(TrackedSocketOptions.LingerState)) linger = LingerState;
+            if (_handle.IsTrackedOption(TrackedSocketOptions.NoDelay)) noDelay = NoDelay;
+            if (_handle.IsTrackedOption(TrackedSocketOptions.ReceiveBufferSize)) receiveSize = ReceiveBufferSize;
+            if (_handle.IsTrackedOption(TrackedSocketOptions.ReceiveTimeout)) receiveTimeout = ReceiveTimeout;
+            if (_handle.IsTrackedOption(TrackedSocketOptions.SendBufferSize)) sendSize = SendBufferSize;
+            if (_handle.IsTrackedOption(TrackedSocketOptions.SendTimeout)) sendTimeout = SendTimeout;
+            if (_handle.IsTrackedOption(TrackedSocketOptions.Ttl)) ttl = Ttl;
```

Comment:
All of these options are properties on Socket, some with a fair amount of non-trivial code behind them.  Getting the value is easy via the property, but doing so accesses the underlying handle, and there can of course be only one handle at a time.  So to do what you propose, I'd need to skip using these properties and instead duplicate all of the code necessary to get/set all of these directly against the handle.